woocommerce/woocommerce
# [Experimental] Surface additional checkout fields values in Admin order page.

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

Integrates additional checkout fields with the order meta box in WP admin. To support this I needed to:

- Add support for `checkbox` inputs in the meta panel
- Add support for passing in `value` and an `update_callback` for fields hooked into the meta box
- Added a hidden input to the checkbox function so it posts `0` when checkboxes are unchecked.
- Add `order` and `context` (edit/view) to the meta panel hooks which allow registration of billing and shipping fields. This is so the correct values and update callbacks can be passed back. This is also so we can show a different value depending on if the values are just being displayed, or being edited via the forms.

Example:

![Screenshot 2024-01-17 at 14 48 13](https://github.com/woocommerce/woocommerce/assets/90977/7cad9594-b64f-4320-aab8-c3dc9c228b12)

![Screenshot 2024-01-17 at 14 48 32](https://github.com/woocommerce/woocommerce/assets/90977/08146b0f-7151-473a-94e1-d2e37aa2c1f5)

Closes #42777 
Closes #42778



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Before testing, implement the snippet to add custom fields as found in #43449 
2. Add an item to cart and place an order via block checkout while filling out the new custom fields
3. Go to the admin dashboard, WooCommerce > Orders and view the order you just placed.
4. In the order details meta box you should see the values you set during checkout. Any checkbox inputs should say "yes" or "no", and select inputs should show the full label e.g. "Do not leave with a neighbor" vs "none".
5. Click the "edit" link beside the billing and shipping sections. The form should toggle on. You should see the new fields in the form.
6. Change some values and "update" the order. Confirm the changes persisted.

To test the changes to the checkbox function:

1. Go to Marketing > Coupons and add/edit a coupon
2. Toggle `Allow free shipping` on and off.
3. Save changes and check changes persist.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

This functionality is behind a feature flag.

</details>
